### ISSUE TITLE
Pull-Down Menus for iOS 14

### ISSUE THREAD

**Issue by YeungKC** [flutter/flutter#60298]
On iOS 14, Pull-Down Menus have been added.

> Menus offer several advantages over action sheets, context menus, and popovers. For example:
>
> A menu opens very near the button that reveals it, so people can instantly understand the relationship between the menu's items and the action they're performing.
>
> In addition to listing actions, a menu can offer selections people can use to affect the primary action.
>
> Menus animate quickly into view and don't dim the screen when they appear, giving both the transition and the overall experience a lightweight feel.

![Image](https://github.com/user-attachments/assets/b8bab64a-2abb-43be-9eee-425aedebe55f)

![Image](https://github.com/user-attachments/assets/834607c4-8ab8-4ded-879e-d921990c1841)

https://developer.apple.com/design/human-interface-guidelines/pull-down-buttons

**Comment by tomgilder** [comment]
These menus are also used as a new way to jump back by long-pressing the navigation bar back button.

This is added by default to iOS 14 apps:

<img src="https://user-images.githubusercontent.com/756862/89769297-8facb000-db05-11ea-8089-5a2557c9c69e.gif" width="300">

**Comment by willlarche** [comment]
API should be similar to the one on Material to make moving between them easy.

**Comment by YeungKC** [comment]
Yes, it might be better to abstract the base from material, and then implement material and cupertino.

**Comment by cpboyd** [comment]
I agree that it'd be nice to have a similar API between Material and Cupertino.
I think that the `CupertinoContextMenu` should also mimic the API of this (or vice-versa) since they're conceptually similar in terms of use (such as occurring during long presses).

The main difference being that the Context Menu blurs everything except the focused item, whereas this leaves everything fully visible.

**Comment by hayderux** [comment]
any updates?

**Comment by Jethro87** [comment]
Any work happening on this?

**Comment by yukitaka13-1110** [comment]
Any update about this?

**Comment by ziqq** [comment]
Any work happening on this?

**Comment by davidhicks980** [comment]
I'm working on a draft implementation with submenus (inspired by @notDmDrl 's [excellent implementation](https://pub.dev/packages/pull_down_button) that he linked above, but with some differences that I'll be interested to get feedback on). I need to do a bit more work before I send in a draft PR, but [here](https://drive.google.com/file/d/15P38NDnE3wvI4FIxZYYUBtIGtXLNAyVb/view?usp=sharing) is a visual preview without some of the fancier layout tricks.

**Comment by ziqq** [comment]
@davidhicks980 Very good implementation. I can't wait to try this.

**Comment by ondanplatform** [comment]
@davidhicks980 Any update about this?

**Comment by davidhicks980** [comment]
I'm just about done. I figured I might as well spend a bit more time on it to make sure I'm not submitting something that is destined to be rejected. I'm on the testing phase... but I think I'll go ahead and submit a draft PR this weekend regardless of how far I get through testing, so as to get feedback from the Flutter team. Thanks for the patience.

**Comment by davidhicks980** [comment]
Hey guys, sorry for not getting a push in yet. I always seem to get caught up in things on the weekends. I pushed changes to this repo if anyone needs a menu ASAP:  https://github.com/davidhicks980/cupertino_menu. I'll get the draft up in a day or two!

**Comment by davidhicks980** [comment]
Whoops -- I just realized my example was broken on the repo above. If anyone tries pulling and running the example, let me know if it looks any different from this: 

https://github.com/flutter/flutter/assets/59215665/746c676c-959f-4d58-bdcc-e777788f189d

**Comment by davidhicks980** [comment]
Pull request is attached! @notDmDrl, would you be okay with signing a CLA? Although our repos are pretty different at this point, your repo is what inspired a lot of what I did. Also, I'll try to figure out how to implement submenus into your repo! I think it's definitely possible, but I'll just need to play with it.

**Comment by notDmDrl** [comment]
> Pull request is attached! @notDmDrl, would you be okay with signing a CLA? Although our repos are pretty different at this point, your repo is what inspired a lot of what I did. Also, I'll try to figure out how to implement submenus into your repo! I think it's definitely possible, but I'll just need to play with it.

Sure 😃 I am currently on a business trip but will sign it as soon as I get back home (next 24 hours).

About those sub-menus in my repo, I am not currently sure whether I want to continue working on the package or simply deprecate it by pointing to your implementation as a replacement. Feel free to create an issue there so we could discuss all the details if you want @davidhicks980 😃

**Comment by kerberjg** [comment]
Thank you all for the hard work!

What is the current status of this issue? Is there a new official Cupertino widget that can be used, or is @notDmDrl's package still the way to go at the current time?

**Comment by sethladd** [comment]
It looks like the updated reference doc is https://developer.apple.com/design/human-interface-guidelines/pull-down-buttons (the doc in the issue's original post is now 404)

**Comment by ercantomac** [comment]
Is anyone currently working on this?

**Comment by davidhicks980** [comment]
@ercantomac yes. Whenever https://github.com/flutter/flutter/pull/158255 lands and a small MenuController PR afterwards, the single-layer [CupertinoMenuAnchor](https://github.com/flutter/flutter/pull/143712) will be shipped. Multi layer will come sometime after that, but it is significantly more complicated.

**Comment by dkwingsmt** [comment]
Status update: @/davidhicks980 has been working on `RawMenuAnchor`, which is a prerequisite for drop down menus. The latest PR https://github.com/flutter/flutter/pull/163481, which allows animations, is still in design discussion phase.

**Comment by davidhicks980** [comment]
Status update: Animations have landed. I'm working on a single layer implementation of a Cupertino menu and menu item. It's fully implemented but I'm only ~50% through testing.

**Comment by dkwingsmt** [comment]
@davidhicks980 Would you like to share your draft PR for an early review?

**Comment by davidhicks980** [comment]
@dkwingsmt Sure thing! A warning though: there are some very rough bits. I'll post it up ASAP
